Classes/Emogrifier.php

```diff
@@ -530,11 +536,21 @@ private function fillStyleAttributesWithMergedStyles()
     private function generateStyleStringFromDeclarationsArrays(array $oldStyles, array $newStyles)
     {
         $combinedStyles = array_merge($oldStyles, $newStyles);
+
+        $cacheKey = serialize($combinedStyles);
+        if (isset($this->caches[self::CACHE_KEY_COMBINED_STYLES][$cacheKey])) {
+            return $this->caches[self::CACHE_KEY_COMBINED_STYLES][$cacheKey];
+        }
+
         $style = '';
         foreach ($combinedStyles as $attributeName => $attributeValue) {
             $style .= (strtolower(trim($attributeName)) . ': ' . trim($attributeValue) . '; ');
         }
-        return trim($style);
+        $trimmedStyle = substr($style, 0, -1);
```

Comment:
The `substr` seems to be marginally faster than the `trim` in my tests. Using `rtrim` isn't any better or worse than `trim`, even when specifying only the space character needs to be trimmed.

I think rewriting the whole thing to use an array and `join` may be faster; typically letting the low-level code deal with things like lots of concatenation is faster in PHP than doing it manually. This also goes for some of the other code.

We're talking about marginal differences in this case, though.